internals/cli/cmd_enter.go

```diff
@@ -139,6 +141,30 @@ func (cmd *cmdEnter) Execute(args []string) error {
 		panic("internal error: expected subcommand (parser.Active == nil)")
 	}
 
+	// The exec command makes the terminal raw to run interactive
+	// applications. Ignore --verbose for those cases, to avoid
+	// log output being sent to the interactive terminal.
+	if execCmd, ok := commander.(*cmdExec); ok {
+		// The following logic is duplicated from cmdExec.Execute()
+		// in internals/cli/cmd_exec.go.
+		stdinIsTerminal := ptyutil.IsTerminal(unix.Stdin)
+		stdoutIsTerminal := ptyutil.IsTerminal(unix.Stdout)
+		var terminal bool
+		if execCmd.Terminal {
+			terminal = true
+		} else if execCmd.NoTerminal {
+			terminal = false
+		} else {
+			terminal = stdoutIsTerminal
+		}
+		// Terminal is made RAW in exec if both terminal and
+		// stdIsTerminal are true.
+		if terminal && stdinIsTerminal {
+			cmd.Verbose = false
+			runCmd.Verbose = false
```

Comment:
A couple of questions:

* Following on from Gustavo's question about logging to FD 2 (stderr), [currently `--verbose` logs to Stdout (FD 1)](https://github.com/canonical/pebble/blob/0a1127ed54ebbd7d89f79450499e0a5907390388/internals/cli/cmd_run.go#L165). Would changing that to Stderr (FD 2) fix this issue, without needing to disable verbose logging?
* We're not disabling other logs here, only "verbose" service logs -- are there other logs printed by `pebble run` or the Pebble CLI that will still cause issues?

Reply:
HI Ben, apologies for the very late reply.

---

> * Following on from Gustavo's question about logging to FD 2 (stderr), [currently `--verbose` logs to Stdout (FD 1)](https://github.com/canonical/pebble/blob/0a1127ed54ebbd7d89f79450499e0a5907390388/internals/cli/cmd_run.go#L165). Would changing that to Stderr (FD 2) fix this issue, without needing to disable verbose logging?

I think, only partially. So yes, if we do change that to `stderr`, we can essentially run the following **in a terminal** without any logs going to vim.

```
pebble enter --verbose --run exec vim 2> logs
```

The problem arises whenever pebble is inside a container and we are running the same stuff with `docker run -it`:

```
docker run -it pebble:test pebble enter --verbose --run exec vim 2> logs
```

There would be nothing in `logs` and the pebble logs would come to the interactive application vim instead. This is because when we use the `-it` flag to docker, our application logs would all come to host's stdout. And since `pebble enter` is specific to containers (rocks), our primary use-case would still have the issue.

---

> * We're not disabling other logs here, only "verbose" service logs -- are there other logs printed by `pebble run` or the Pebble CLI that will still cause issues?

For `pebble run`, the impact of `-v`/`--verbose` option is only the addition of the service outputs as you linked above. There will still be other logs without `--verbose` like below:

```
2024-02-19T09:26:13.226Z [pebble] Started daemon.
2024-02-19T09:26:13.277Z [pebble] POST /v1/services 49.446134ms 202
2024-02-19T09:26:13.277Z [pebble] Started default services with change 359.
...
```

For `pebble enter exec` however, if there is no `--verbose` specified, the other logs will not be printed at all. 

https://github.com/canonical/pebble/blob/0a1127ed54ebbd7d89f79450499e0a5907390388/internals/cli/cmd_enter.go#L164-L166

So if we disable `--verbose`, there will be no logs altogether that may cause any issues for interactive exec runs.